BoyumFoosballStats/Services/MatchMakingService.cs

```diff
@@ -24,6 +24,84 @@ public MatchMakingService(IAzureBlobStorageHelper blobStorageHelper)
+            return await FindFairestMatch(players, matchMakingMethod ?? MatchMakingMethod.Ai);
+        }
+
+        var lastMatch = matches.Last();
+        var newMatch = new Match
+        {
+            BlackAttackerPlayer = lastMatch.BlackAttackerPlayer,
+            BlackDefenderPlayer = lastMatch.BlackDefenderPlayer,
+            GreyAttackerPlayer = lastMatch.GreyAttackerPlayer,
+            GreyDefenderPlayer = lastMatch.GreyDefenderPlayer
+        };
+        var playersInLastMatch = players.Where(x => lastMatch.Players.Any(y => y.Id == x.Id)).ToList();
+        var playersNotInLastMatch = players.Except(playersInLastMatch).ToList();
+        if (playersNotInLastMatch.Count >= 4)
+        {
+            //If all players need to swap - find the fairest match
+            return await FindFairestMatch(playersNotInLastMatch, matchMakingMethod ?? MatchMakingMethod.Ai);
+        }
+
+        var matchesPlayedByPlayer = playersInLastMatch
+            .ToDictionary(
+                player => player,
+                player => matches
+                    .OrderByDescending(match => match.MatchDate)
+                    .TakeWhile(match => match.Players.Any(x => x.Id == player.Id))
+                    .Count()
+            )
+            .Where(kv => kv.Value > 0)
+            .ToDictionary(kv => kv.Key, kv => kv.Value);
+        
+        if (matchMakingMethod != null)
+        {
+            //If a matchmaking method was explicitly specified - find the fairest match
+            var playersToKeep = matchesPlayedByPlayer.OrderBy(x => x.Value).Take(4 - playersNotInLastMatch.Count);
+            var playersForNewMatch = playersToKeep.Select(x => x.Key).ToList();
+            playersForNewMatch.AddRange(playersNotInLastMatch);
+            return await FindFairestMatch(playersForNewMatch, matchMakingMethod.Value);
+        }
+
+        //If none of the other cases are true - swap each player individually
+        var playersToSwap = matchesPlayedByPlayer.OrderByDescending(x => x.Value)
+            .ThenBy(x => lastMatch.Winners.Contains(x.Key)).Take(playersNotInLastMatch.Count);
+        var swapIndex = 0;
+        foreach (var playerToSwap in playersToSwap)
+        {
+            var swapId = playerToSwap.Key.Id;
+            if (lastMatch.BlackAttackerPlayer?.Id == swapId)
+            {
+                newMatch.BlackAttackerPlayer = playersNotInLastMatch[swapIndex];
+            }
+
+            if (lastMatch.BlackDefenderPlayer?.Id == swapId)
+            {
+                newMatch.BlackDefenderPlayer = playersNotInLastMatch[swapIndex];
+            }
+
+            if (lastMatch.GreyAttackerPlayer?.Id == swapId)
+            {
+                newMatch.GreyAttackerPlayer = playersNotInLastMatch[swapIndex];
+            }
+
+            if (lastMatch.GreyDefenderPlayer?.Id == swapId)
+            {
+                newMatch.GreyDefenderPlayer = playersNotInLastMatch[swapIndex];
+            }
```

Comment:
Can't you use the new fancy `Players` property that you've added? it will simplify this logic quite a bit

Reply:
Not unless I'm missing something - the problem is that I have to figure out which exact spot the player was in, not just whether they were in the match.

I did consider creating an extension method to Swap(oldPlayer, newPlayer) to wrap the logic elsewhere, but I'm not sure it would be used elsewhere.